mzLib/MassSpectrometry/PeakIndexing/MassIndexingEngine.cs

```diff
@@ -0,0 +1,138 @@
+﻿using MathNet.Numerics.RootFinding;
+using MzLibUtil;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MassSpectrometry
+{
+    public class MassIndexingEngine: IndexingEngine <IndexedMass>
+    {
+        protected override int BinsPerDalton => 1;
+        public int MaxMass { get; set; } = 30000;
+        public ScanInfo[] ScanInfoArray { get; private set; }
+
+        public MassIndexingEngine()
+        {
+        }
+
+        public bool IndexPeaks(MsDataScan[] scanArray, DeconvolutionParameters deconParameters, MzRange mzRange = null, double minMass = 0, int minCharge = 1)
+        {
+            if (scanArray.IsNullOrEmpty() || scanArray.All(p => p == null))
+                return false;
+
+            IndexedPeaks = new List<IndexedMass> [MaxMass];
+            ScanInfoArray = new ScanInfo[scanArray.Length];
+            for (int scanIndex = 0; scanIndex < scanArray.Length; scanIndex++)
+            {
+                ScanInfoArray[scanIndex] = new ScanInfo(scanArray[scanIndex].OneBasedScanNumber, scanIndex, scanArray[scanIndex].RetentionTime);
+                var envelopes = Deconvoluter.Deconvolute(scanArray[scanIndex].MassSpectrum, deconParameters, mzRange);
+                foreach (var envelope in envelopes)
+                {
+                    if (envelope.MonoisotopicMass < minMass || envelope.Charge < minCharge)
+                        continue;
+                    int roundedMass = (int)Math.Round(envelope.MonoisotopicMass * BinsPerDalton, 0);
+                    IndexedPeaks[roundedMass] ??= new List<IndexedMass>();
+                    IndexedPeaks[roundedMass].Add(new IndexedMass(envelope, scanArray[scanIndex].RetentionTime, scanIndex, scanArray[scanIndex].MsnOrder));
+                }
+            }
+            if (IndexedPeaks == null || IndexedPeaks.Length == 0)
+                return false;
+            else
+                return true;
+        }
+
+        public IndexedMass GetIndexedPeak(double mass, int charge, int zeroBasedScanIndex, PpmTolerance ppmTolerance)
+        {
+            var bins = GetBinsInRange(mass, ppmTolerance);
+            if (bins.Count == 0) return null;
+            List<int> peakIndicesInBins = bins.Select(b => BinarySearchForIndexedPeak(b, zeroBasedScanIndex)).ToList();
+            return GetBestPeakFromBins(bins, mass, charge, zeroBasedScanIndex, peakIndicesInBins, ppmTolerance);
+        }
+
+        public static IndexedMass GetBestPeakFromBins(List<List<IndexedMass>> allBins, double mass, int charge, int zeroBasedScanIndex, IList<int> peakIndicesInBins, PpmTolerance ppmTolerance)
+        {
+            IndexedMass bestMass = null;
+            for (int i = 0; i < allBins.Count; i++)
+            {
+                var tempPeak = GetPeakFromBin(allBins[i], mass, charge, zeroBasedScanIndex, peakIndicesInBins[i], ppmTolerance);
```

Comment:
It looks like you're iterating over all bins here, which is inefficient. you should select the bins that are in tolerance